### ISSUE TITLE
[BUG] Longhorn manager deleted all backup from the backuptarget during the re-installation

### ISSUE THREAD

**Issue by itsmanikandanm** [longhorn/longhorn#9384]
## Describe the bug

I have `NFS` backup target and i am working for disaster recovery docs. When i uninstall longhorn and reinstall with same backup target its wiped all volumes from the backup target.

## Found this from `longhorn-manager` logs

```console
level=info msg="Complete deleting backup volume pvc-65fb00 │
│ 93-bf0f-49b4-8fab-8f37081e0d5a" func="engineapi.(*BackupTargetClient).BackupVolumeDelete" file="backups.go:2 │
│ 28"          
```

## Found this from `longhorn-manager`  repo

```console
// BackupVolumeDelete deletes the backup volume from the remote backup target
func (btc *BackupTargetClient) BackupVolumeDelete(destURL, volumeName string, credential map[string]string) error {
	_, err := btc.ExecuteEngineBinaryWithoutTimeout("backup", "rm", "--volume", volumeName, btc.URL)
	if err != nil {
		if types.ErrorIsNotFound(err) {
			return nil
		}
		return errors.Wrapf(err, "error deleting backup volume")
	}
	logrus.Infof("Complete deleting backup volume %s", volumeName)
	return nil
}          
```

## Environment

 - Longhorn version:  `1.7.0`
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): `Fluxcd and Helm`
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: `RKE2`
   - Number of control plane nodes in the cluster: `Single Node Cluster`
   - Number of worker nodes in the cluster: `Single Node Cluster`
 - Node config
   - OS type and version: `Linux and Ubuntu 22`

**Comment by ejweber** [comment]
A couple of questions come to mind quickly:

1. Are the backups deleted from the target during the uninstall, or during/after the reinstall? To check, take a look at the files in the NFS export before doing the uninstall, after doing the uninstall, and after doing the reinstall. During what period do files disappear?
2. How is the uninstall done? Helm? kubectl? Please describe the method you use in as much detail as you can.

**Comment by itsmanikandanm** [comment]
Hi @ejweber,

Thank you for your reply. 

1. During the uninstall backup available on the 'NFS' server. But when i reinstall the longhorn, 'longhorn-manager' initialise backup target that time backup has been wiped by the manager. 

2. During the reinstall I got these logs from manager pod. 

3. Uninstall complete by helm with fluxcd, after that I will cleanup CRDs with 'for loop' command which is mentioned in the uninstall docs. 

I cleared manually if namespace stuck at terminating state. But it was removed by my fluxcd.

**Comment by ejweber** [comment]
The for loop command you are referring to is only necessary to clean up very broken uninstallations. It is generally not necessary, and if it is actually doing anything in your workflow (i.e. it actually finds resources to clean up), it may be an indicator of an uninstall problem.

From some quick experimentation, it looks like the command like `flux delete helmrelease longhorn-release -n longhorn-system` returns immediately, even though the underlying Longhorn uninstall is still ongoing. Do you wait until all Longhorn pods have shutdown (check with `kubectl get pod -n longhorn-system`) before proceeding to the for loop or doing the reinstall?

Is it possible for you to upload a support bundle containing the logs (after the reinstall) to this ticket or send one to longorn-support-bundle@suse.com?

**Comment by itsmanikandanm** [comment]
Hi @ejweber,

1. I am uninstalling longhorn using command like this `flux delete helmrelease longhorn-helm-release -n longhorn-system` or simply `comment out` helmrelease in my repository.

2. Once flux done the uninstall, i will check pod termination status. once all the pod terminated then i will run below for loop command to delete the CRD only.

```sh
for crd in $(kubectl get crd -o jsonpath={.items[*].metadata.name} | tr ' ' '\n' | grep longhorn.io); do
  kubectl delete crd/$crd
  done
```

If any CRD is stuck at deleting state then i will run below command to fix that

```sh
kubectl patch customresourcedefinitions backups.longhorn.io -p '{"metadata":{"finalizers":null}}' --type=merge
kubectl patch customresourcedefinitions backupvolumes.longhorn.io -p '{"metadata":{"finalizers":null}}' --type=merge
kubectl patch customresourcedefinitions nodes.longhorn.io -p '{"metadata":{"finalizers":null}}' --type=merge
kubectl patch customresourcedefinitions engineimages.longhorn.io -p '{"metadata":{"finalizers":null}}' --type=merge
```

4.Once all the resources deleted on the `longhorn-system` namespace, i will `flux delete kustomization longhorn`. Then the flux will clear all components which is created by the flux during the installation.

5. Finally, go to `NFS` server and check the backup volumes, then i will find output like this

```console
backupstore/
└── volumes
    ├── 88
    │   └── 56
    │       └── pvc-77b8e56a-b2c4-486b-b1e0-9b9a74161510
    │           ├── backups
    │           │   └── backup_backup-d7db030c575d4fbe.cfg
    │           ├── blocks
    │           │   ├── 0e
    │           │   │   └── 79
    │           │   │       └── 0e79eb11677369f35e26a4febc82f9fcf3fb9d26336bc53d037435aeb0e017a2.blk
    │           │   ├── 19
    │           │   │   └── 35
    │           │   │       └── 1935bf07ec7f56399904c0b46e53fabcbe36e9804a3beab68ab60cec6644cca3.blk
    │           │   ├── 1c
    │           │   │   └── 5a
    │           │   │       └── 1c5a229f404c25b2279dfe2388de0ece245bed62f8e59a0ab07fe74e414ba3c2.blk
    │           │   ├── 24
    │           │   │   └── 2c
    │           │   │       └── 242c1d20f0f0873bfc0dbd491543f276314f7b4d3e58516f37c052596d940674.blk
    │           │   ├── 2b
    │           │   │   ├── 0f
    │           │   │   │   └── 2b0faf00a2627d62560d886a08a9f6bfb8a47e04889fd207d378a64d0b160f11.blk
    │           │   │   └── 15
    │           │   │       └── 2b1597ffd3c81c4ee3362ee918bb3d653e6663f45ac8a24b56db4bad00b9b70a.blk
    │           │   ├── 2c
    │           │   │   ├── 38
    │           │   │   │   └── 2c38c61b2def49fb3aa5fffada76fc6d5a236aafe480790abd39615217bdba32.blk
    │           │   │   └── ca
    │           │   │       └── 2ccad3b778990ea8f8d3ff45b1bd0c39b7f9f87ee72740c03b9f0df1a8d0bd02.blk
    │           │   ├── 2d
    │           │   │   └── 03
    │           │   │       └── 2d030529aebaa238aa279158f5370ecc9861bee94947bce82cb5b26980c608f7.blk
    │           │   ├── 2f
    │           │   │   └── 68
    │           │   │       └── 2f6827921fa568e965028800aa9a16b60f139fca6294798ee55a1ce7bb8c8570.blk
    │           │   ├── 38
    │           │   │   └── 45
    │           │   │       └── 38457dcb8f8bceade8aea5a51eda6a62ca63d2d7ffad25426b7c641b218f3ede.blk
    │           │   ├── 3d
    │           │   │   └── c5
    │           │   │       └── 3dc58a1249e83a2f67f3bb3668f10ae2c49bc11514ee44598703b946a9c437f6.blk
    │           │   ├── 3f
    │           │   │   └── f5
    │           │   │       └── 3ff5da1899edae8b90f347cb99d22e258ed689c606f320ab3bfc7ba65e28c8bf.blk
    │           │   ├── 45
    │           │   │   └── 2f
    │           │   │       └── 452f226de1baf265b8b57a32aa9998692af8d28b7fcc40fde14d74cce7bd5a04.blk
    │           │   ├── 49
    │           │   │   ├── 65
    │           │   │   │   └── 496524c4ca973a9a6275237e518e4cf89a5483ef5cea6bc45c5a4aebac49bed3.blk
    │           │   │   └── f3
    │           │   │       └── 49f33caaebea30b9d92a8d6c2c7dc16cd3b49ca7309121bf8d1352cd61fc73d8.blk
    │           │   ├── 4d
    │           │   │   └── 72
    │           │   │       └── 4d7213e054be0e1fd836618a729a01399a938b9ffd8c863ae838c327452f6563.blk
    │           │   ├── 55
    │           │   │   └── 18
    │           │   │       └── 5518d624d951576d40642c422e2bd6f06a74adb3dfa0b489cbcc26422498b551.blk
    │           │   ├── 59
    │           │   │   └── c7
    │           │   │       └── 59c778a4882d3efe6b87f92330422a7ea8ab01d009dbe5f50c729bd5ae46300e.blk
    │           │   ├── 64
    │           │   │   ├── 0d
    │           │   │   │   └── 640d5704edb7f149b21e9b87d1803f7ca47d241793965e3edb98ad653684f40f.blk
    │           │   │   └── 3b
    │           │   │       └── 643b3db7412d7cb73db136b4e33270b47ebcee7a1f613f96402d0721626ecac9.blk
    │           │   ├── 68
    │           │   │   ├── 0b
    │           │   │   │   └── 680bb70fad1886906a198f3a99d0ae354328b136adbdf6db25f514b973aade7b.blk
    │           │   │   └── 2c
    │           │   │       └── 682ccdf47121155fc90e4b77c394929f8d9157d9a42afb753883d1633f79826f.blk
    │           │   ├── 6d
    │           │   │   └── bd
    │           │   │       └── 6dbdefbf8ef37e5adaae539f1d7e76bcc4eb62b656a468bf065d115f6ade27e2.blk
    │           │   ├── 6e
    │           │   │   └── 43
    │           │   │       └── 6e43ed2a81602cce0601d80791c04ec7a9cf8d9476198caa90feaaab4772120c.blk
    │           │   ├── 70
    │           │   │   └── 31
    │           │   │       └── 703107248d513f700a4b7b3cd0e4a28aa76eff0f49cb70e6f4f00514fd3d9f6f.blk
    │           │   ├── 73
    │           │   │   └── 18
    │           │   │       └── 731859029215873fdac1c9f2f8bd25a334abf0f3a9e1b057cf2cacc2826d86b0.blk
    │           │   ├── 7a
    │           │   │   └── a4
    │           │   │       └── 7aa44d30fa2e125ed25d7db1b1377b91b0fcfab7fdaa86f6d8cd359f9a956fdc.blk
    │           │   ├── 7b
    │           │   │   └── ef
    │           │   │       └── 7bef31764ad6c490161a9c918790ba7f639f1bc4b5f5059239836110613b387e.blk
    │           │   ├── 7f
    │           │   │   └── d7
    │           │   │       └── 7fd7d76bc80eefa1118aed0558fee03c31ca846d29404d3516b4fe9597615d37.blk
    │           │   ├── 87
    │           │   │   ├── 91
    │           │   │   │   └── 8791bd0b1e550cfcef32508883b86ba0f4fbd7bdd976b52031b8e8a7eaeecb26.blk
    │           │   │   └── 99
    │           │   │       └── 8799a77a6342ae2a4e35a70637ca9227aa1de643fc81340547ad14e5746cbe19.blk
    │           │   ├── 90
    │           │   │   ├── 16
    │           │   │   │   └── 9016537e389b1ebe3b0375e45e9f201a5f4626a9fd4c8bbb04c054fd03eaad60.blk
    │           │   │   └── 26
    │           │   │       └── 90263874e47af31b71a2df72d647976f248624318f51b94d8f839f9a9de24935.blk
    │           │   ├── 92
    │           │   │   └── 17
    │           │   │       └── 92171af337d9a2505a20c057d8901f966e9406473f43afc4d8d0f9cc00de4b85.blk
    │           │   ├── 97
    │           │   │   ├── ad
    │           │   │   │   └── 97ad8f60ed5a4c013bc247c54f60a88ca072cbbf40be3e14cfadaf485b9f5518.blk
    │           │   │   └── ff
    │           │   │       └── 97ff025f55465fcd30716ae533dd274472be4a2160bc38aa8611513d62369bb5.blk
    │           │   ├── 99
    │           │   │   └── 65
    │           │   │       └── 99650e787254f9f323aede41efb49e4a96f140bcbcd07e2ee661d6a8bca64c51.blk
    │           │   ├── 9d
    │           │   │   └── 07
    │           │   │       └── 9d077b775bf0da6df51212598584580fa2a5ff161009e03176916099624bfe00.blk
    │           │   ├── 9f
    │           │   │   └── 5f
    │           │   │       └── 9f5f172ca9e8ec2bacc8821ce731e79b307524c72f04d62e35ba4b17ec6e5c33.blk
    │           │   ├── a5
    │           │   │   └── 7f
    │           │   │       └── a57f4a7b3e4fb1448d67a6154e619c0f80adc7014e7a4a0ba09835dd55ec40cf.blk
    │           │   ├── a7
    │           │   │   ├── 10
    │           │   │   │   └── a7106078d2e278ffc3da12804fa991aed9e60ac63ce24b97aaabf0c153ba3598.blk
    │           │   │   └── 1a
    │           │   │       └── a71a057020ffe4446c9b566615a38e8325ec0f9abf32df605ffb7aabad0723ad.blk
    │           │   ├── aa
    │           │   │   └── 68
    │           │   │       └── aa6856a556cd0e44e6586faa9e9c2e01e570d88ad64325403a92943f103ed8be.blk
    │           │   ├── b8
    │           │   │   └── f0
    │           │   │       └── b8f01739a28527e159363d514d7632f08c8c27a287109acb4f9c6234e94eba22.blk
    │           │   ├── bc
    │           │   │   └── 6b
    │           │   │       └── bc6bc12ccd6f86fb44ce17b7b65dddca54ef5bfcb0962df58ce67c8062388fcb.blk
    │           │   ├── c2
    │           │   │   └── 9d
    │           │   │       └── c29d537206d9673f44c9795f110924c725f2ef284236793d65664d5c0661a5a2.blk
    │           │   ├── c9
    │           │   │   └── da
    │           │   │       └── c9da2d3b977343349b3c34169143d419537591b8cfffd49e874c4fd2cbd68f82.blk
    │           │   ├── cf
    │           │   │   └── 7b
    │           │   │       └── cf7b1244319195bf8e0104eb836a101a2b1cb57f3a4fd81d9008abf7cb61fa13.blk
    │           │   ├── d1
    │           │   │   └── c9
    │           │   │       └── d1c92abd3f7844819bb5bef84d7ffe4cac0314c9bc527b39259fa944b0901e73.blk
    │           │   ├── da
    │           │   │   └── 6d
    │           │   │       └── da6dedd50a280181a0ee0a08078d500c2e66e344044e486aa363a58609f637b5.blk
    │           │   ├── e7
    │           │   │   └── 52
    │           │   │       └── e7522afe766b125f135f75771206173ede4398c0709c361cc84a95abcb58e928.blk
    │           │   ├── eb
    │           │   │   ├── 18
    │           │   │   │   └── eb18c3a5bc5b559178066d84e5819ddc7f2f749cb6476bd0b9bedfab7eba3c39.blk
    │           │   │   └── 41
    │           │   │       └── eb4193a88d9cefe74abd7c0124e654d45ce62fdf49bf29d7cf6f7b6b935191f2.blk
    │           │   ├── ef
    │           │   │   └── 12
    │           │   │       └── ef12b23589336c66eba3c4c52be65c1955469f8891c1f27ba688e94c5759385f.blk
    │           │   ├── f1
    │           │   │   └── 4a
    │           │   │       └── f14ab888344335e033e7a3442799cc9c0ac00ef5db2d3cc38df0d2e4d7ab3cf0.blk
    │           │   └── f3
    │           │       └── 1b
    │           │           └── f31bcb77f721d5a22fb7036de56d836fe521e641acab6d55df655cc0b0748ced.blk
    │           ├── locks
    │           │   └── lock-2f35e28854b54cac.lck
    │           └── volume.cfg
    ├── dc
    │   └── b4
    │       └── pvc-dc677b2d-5a97-493d-b8f1-cc1f1e6d2814
    │           ├── backups
    │           │   └── backup_backup-a1f110d036bb47b2.cfg
    │           ├── blocks
    │           │   ├── 1a
    │           │   │   └── e7
    │           │   │       └── 1ae7cd7bb0d99c7941c0f912d71c00620698bca2c410a7622a41228ff1fe14fc.blk
    │           │   ├── 24
    │           │   │   └── f7
    │           │   │       └── 24f79783c7b05b936b70b66e46d2a093d9b7fc33689dddbe06ec6caf20444572.blk
    │           │   ├── 2d
    │           │   │   └── 48
    │           │   │       └── 2d48132f298d05f7010708b1dc9c659ef99521b8795bec53719ce74cd02e8240.blk
    │           │   ├── 44
    │           │   │   └── 2f
    │           │   │       └── 442f6a1c7235d4a74c1cefd641ee125fca9b78c34784d672413ed4cdf2593079.blk
    │           │   ├── 67
    │           │   │   └── 6c
    │           │   │       └── 676cb45917c6647d67077f670f4600133335094d0d23a1742368fbfb1ded0bb3.blk
    │           │   ├── 73
    │           │   │   └── 18
    │           │   │       └── 731859029215873fdac1c9f2f8bd25a334abf0f3a9e1b057cf2cacc2826d86b0.blk
    │           │   ├── 87
    │           │   │   └── ce
    │           │   │       └── 87ce124386a39f34627244964454c9439a41515c3bf8711109b3d7cbd6e0e170.blk
    │           │   ├── 90
    │           │   │   └── 26
    │           │   │       └── 90263874e47af31b71a2df72d647976f248624318f51b94d8f839f9a9de24935.blk
    │           │   ├── 9b
    │           │   │   └── 10
    │           │   │       └── 9b10d3b8c65dda277ff28aa14ce01f7e54d87c2612c69965926a3d0adc2d8c1e.blk
    │           │   ├── a9
    │           │   │   └── da
    │           │   │       └── a9dac41ac36dabbbaf84203e6f14cc5c28f11a9dd8a73628392db1b90676cca1.blk
    │           │   ├── b9
    │           │   │   └── a5
    │           │   │       └── b9a5e1662452555a67d51c42b73d63087a5ceca3a584b530d32376fc28ea5518.blk
    │           │   ├── bb
    │           │   │   └── 63
    │           │   │       └── bb63b51bb8fdb425632493c20414bedca4c17e94b218e922f7e0fb0945491c02.blk
    │           │   ├── c4
    │           │   │   └── 13
    │           │   │       └── c41380860e348fb200b04d37ad3a72e3ec8ca36dba82fcec3e6672179f13c351.blk
    │           │   ├── d7
    │           │   │   └── d2
    │           │   │       └── d7d26a26a5e255b57dffda8ce496133c27dac1ef2f3a90b86fff15b949513111.blk
    │           │   └── d8
    │           │       └── ba
    │           │           └── d8bae55ca497f712cd8d4cf94a14023f0e1c83da68f80c9833f8417929d2569d.blk
    │           ├── locks
    │           │   └── lock-b39b6c617abd42f1.lck
    │           └── volume.cfg
    └── ed
        └── 00
            └── pvc-8b6d2bca-8eb3-405c-93ec-a4c26acad64c
                ├── backups
                │   └── backup_backup-d310929023384a05.cfg
                ├── blocks
                │   ├── 33
                │   │   └── 0a
                │   │       └── 330a548bf0baa276fc354ad03bf21f21248b1bdcebcd98ca9ae334055f328ff8.blk
                │   ├── 4a
                │   │   └── 8c
                │   │       └── 4a8cf850f39d2fcc4abf56b344a316b350c4f636d1814a736412f240ddca2295.blk
                │   ├── 50
                │   │   └── 6a
                │   │       └── 506a41a6bffbffdea0fc364129b86b83eb793243b3809582d3eae4944bee1b3a.blk
                │   ├── 5e
                │   │   └── 41
                │   │       └── 5e41a73f0670908b6daa4c520410f8da35e3c936873be2d659afc86a53efcdfd.blk
                │   ├── 72
                │   │   └── d2
                │   │       └── 72d2bcafb9216b9842e8cbfa8e47e69920f60da78dc9074247018af4657aaff5.blk
                │   ├── 73
                │   │   └── 18
                │   │       └── 731859029215873fdac1c9f2f8bd25a334abf0f3a9e1b057cf2cacc2826d86b0.blk
                │   ├── 76
                │   │   ├── 6b
                │   │   │   └── 766b0a44e6a613f91099e5eb6633dcaf279610f09b2966d8c0bcc17e667e596b.blk
                │   │   └── e6
                │   │       └── 76e681255e17860937b7c4ac37db625e2df571899c3251e4785d544985a5ba16.blk
                │   ├── 8e
                │   │   └── b9
                │   │       └── 8eb9c06656906c0d448315690d1546c7a768ddff866bff134ff948227897089f.blk
                │   ├── 90
                │   │   └── 26
                │   │       └── 90263874e47af31b71a2df72d647976f248624318f51b94d8f839f9a9de24935.blk
                │   ├── b4
                │   │   └── a3
                │   │       └── b4a358b57997a9bfbd5129ebd72f02e5c5230c7854317148b2a43751819b1dbe.blk
                │   ├── d3
                │   │   └── 5a
                │   │       └── d35af474cf14012d553aa38377c2f3d5aad8159efbec74aff494af8f48eab919.blk
                │   ├── d8
                │   │   └── ba
                │   │       └── d8bae55ca497f712cd8d4cf94a14023f0e1c83da68f80c9833f8417929d2569d.blk
                │   ├── e1
                │   │   └── f1
                │   │       └── e1f120365645a0173af4f39c2f6532e2bde942904e7cbfc686d999e4cf2131ed.blk
                │   └── ee
                │       └── 2a
                │           └── ee2ab2d8170b63dcf04c59039af5a89e9c49d6bfde442afb9ae173870d26c564.blk
                ├── locks
                │   └── lock-4fa203bab46b4986.lck
                └── volume.cfg
``` 

6. Once reinstalled with flux like previous, then i got like this

```console
root@server:/mnt/k8s_backup#tree backupstore/
backupstore/

0 directories, 0 files

```

7. I will attach support bundle then let you know. Meanwhile please check `longhorn-manager` pod log which i have attached for you reference.
[longhorn-system-longhorn-manager-jkfkj-1725528608839622635.log](https://github.com/user-attachments/files/16887242/longhorn-system-longhorn-manager-jkfkj-1725528608839622635.log)

**Comment by ejweber** [comment]
I combed through the provided log and code, but I think we will need the support bundle to investigate further.

It seems we only delete the data for a BackupVolume from the backup store IF:

- The BackupVolume has a deletion timestamp, and
- the BackupTarget.Spec.BackupTargetURL is set.

https://github.com/longhorn/longhorn-manager/blob/38a74bb562208d4617d80f2cc9898b4b4fbcbe63/controller/backup_volume_controller.go#L224-L245

It seems the BackupTarget.Spec.BackupTargetURL must be correctly set early, since we start the sync timer.

```
longhorn-manager time="2024-09-05T14:57:32+05:30" level=info msg="Starting backup store timer" func="controller.(*BackupStoreTimer).Start" file="setting_controller.go:1282" component=backup-store-timer controller=longhorn-setting interval=5m0s node=master-01
```

Almost a minute later, it MUST be the case the above conditions must be met, since we log:

```
longhorn-manager time="2024-09-05T14:58:28+05:30" level=info msg="Complete deleting backup volume pvc-dc677b2d-5a97-493d-b8f1-cc1f1e6d2814" func="engineapi.(*BackupTargetClient).BackupVolumeDelete" file="backups.go:228"
...
```

So either an external component deleted the BackupVolumes a minute later, or Longhorn did at:

- https://github.com/longhorn/longhorn-manager/blob/38a74bb562208d4617d80f2cc9898b4b4fbcbe63/controller/backup_target_controller.go#L494-L506
- https://github.com/longhorn/longhorn-manager/blob/38a74bb562208d4617d80f2cc9898b4b4fbcbe63/controller/backup_target_controller.go#L653-L671

I can see no obvious path for Longhorn to have done it at those locations, but I am hoping the support bundle can provide more context around timing (it will contain YAML representations of all Longhorn resources), etc.

By the way, is this reproducible every time on your end @itsmanikandanm? Neither @derekbit or I have been able to reproduce it yet.

**Comment by itsmanikandanm** [comment]
Hi @ejweber,

Thank you for your time to investicate my issues.

I conducted R&D to identify the root cause of the issue: executing `flux delete kustomization longhorn` and forcefully clearing CRDs. Subsequently, I reinstalled the chart, then `longhorn-manager`, which wiped the backup target during the `instance-manager` setup.

Uninstalling the chart with `flux delete helmrelease longhorn` or `helm uninstall longhorn -n longhorn-system` won't wipe anything upon re-installation, Also existing backups can be adopted by Longhorn.

Once again thank you for your valuable time.
